Lua on each mapgen thread (#13092)
minetest/minetest: doc/lua_api.md

### `gennotify`

Returns a table. You need to announce your interest in a specific
field by calling `minetest.set_gen_notify()` *before* map generation happens.

* key = string: generation notification type

> * value = list of positions (generally)

Reviewer: ```suggestion
* value = list of positions (usually)
```

Returns a table mapping requested generation notification types to arrays of
positions at which the corresponding generated structures are located within
the current chunk.

For this to contain anything you need to announce your interest in a specific
field by calling `minetest.set_gen_notify()` *before* map generation happens.

> Possible fields of the returned table are:

Author: I will re-do the section as described but I don't like the "set_gen_notify must be used to _specify_" wording.

* `minetest.register_on_punchnode(function(pos, node, puncher, pointed_thing))`
    * Called when a node is punched
* `minetest.register_on_generated(function(minp, maxp, blockseed))`
    * Called after generating a piece of world between `minp` and `maxp`.

>       Modifying nodes inside the area is a bit faster than usual.

Author: I think this is actually referring to how `MapEditEvent`s do not need to be processed or sent to clients. So no relation.

>       Modifying nodes inside the area is a bit faster than usual.

Reviewer 2: If we're left guessing then it's probably best to either reword or remove this line due to obscurity.

* `minetest.get_voxel_manip([pos1, pos2])`
    * Return voxel manipulator object.
    * Loads the manipulator from the map if positions are passed.
* `minetest.set_gen_notify(flags, deco_ids, custom_ids)`
    * Set the types of on-generate notifications that should be collected.
    * `flags` is a flag field with the available flags:
        * dungeon, template, cave_begin, cave_end, large_cave_begin

>         * large_cave_end, decoration, custom

Reviewer 2: Single source of truth. Link not tested.
```
    * `flags`: flag field, see [`gennotify`] for available generaton notification types.
```

    * with all functions and userdata values replaced by `true`, calling any
      callbacks here is obviously not possible

Mapgen environment
------------------

The engine runs the map generator on separate threads, each of these also has
a Lua environment. Its primary purpose is to allow mods to operate on newly
generated parts of the map to e.g. generate custom structures.
Internally it is referred to as "emerge environment".

Refer to the above section for the usual disclaimer on what environment isolation entails.

> * `minetest.register_mapgen_dofile(path)`:

Author: hmm I agree that's much better. `register_async_dofile` should be renamed too ideally

    * Register a path to a Lua file to be imported when a mapgen environment
      is initialized. You'd put your mapgen-related code in there.

### List of APIs exclusive to the mapgen env

* `minetest.register_on_generated(function(vmanip, minp, maxp, blockseed))`
    * Called after the engine mapgen finishes a chunk.
      The chunk data resides in `vmanip`. Other parts of the map are not accessible.
      The area of the chunk if comprised of `minp` and `maxp`, note that is smaller
      than the emerged area of the VoxelManip.

>       Note that you don't need to call `read_from_map()` or `write_to_map()` on
>       the VoxelManipulator, doing so is in fact disallowed.

Reviewer 2: Based on the deprecation message in the code I would rewrite it as follows

```suggestion
      Note: calling `read_from_map()` or `write_to_map()` on the VoxelManipulator object
      is not necessary and is disallowed.
```

    * `blockseed`: 64-bit seed number used for this chunk
* `minetest.save_gen_notify(id, data)`

>     * Saves data for retrieval using the gennotify mechanism (see [Mapgen objects]).
>     * Data is bound to the chunk that is currently being processed, so this function
>       only makes sense inside the above-mentioned generation callback.

Reviewer 2: ```suggestion
    * Used to transfer data specific to the currently processed chunk across different mapgen envs (see [Mapgen objects]).
    * (execution order?)
 ```
It is yet no clear to me which `register_on_generated` callback is executed first - the one of the main thread or the one inside the mapgen env.  That would need clarification in the introduction text above and perhaps referenced here.

Reviewer 2: Turns out I did not understand the concept of this function. What does it do? How is it helpful in mapgen envs?

Reviewer 2: Does that mean that the separate mapgen envs are executed first so that the gen notifications can be populated? Otherwise I fail to understand how this information could be obtained beforehand.
I also do not see any mention of the execution order.

Author: emerge scripting is only initialized after mods load.
at that point the wanted mapgen, flags, noise settings, etc. and also the set of wanted gennotify data is known.

>  I also do not see any mention of the execution order.

You mean in case of multiple `register_mapgen_script` or what else?

    * `id`: user-defined ID (a string)
      By convention these should be the mod name with an optional
      colon and specifier added, e.g. `"default"` or `"default:dungeon_loot"`

>     * `data`: any Lua object (will be serialized, no userdata allowed)

Reviewer 2: +
```
    * Returns `true` on success
```

Author: If you use this function according to the documentation there is no way it can fail.
Should I still document this?

### List of APIs available in the mapgen env

Classes:
* `AreaStore`
* `ItemStack`
* `PerlinNoise`
* `PerlinNoiseMap`
* `PseudoRandom`
* `PcgRandom`
* `SecureRandom`
* `VoxelArea`
* `VoxelManip`
    * only given by callbacks; no access to rest of map

> * `Settings`

Author: Settings: thread-safe and read write
ItemStack: only operates on local data

    * with all functions and userdata values replaced by `true`, calling any
      callbacks here is obviously not possible

Mapgen environment
------------------

The engine runs the map generator on separate threads, each of these also has
a Lua environment. Its primary purpose is to allow mods to operate on newly
generated parts of the map to e.g. generate custom structures.
Internally it is referred to as "emerge environment".

Refer to [Async environment] for the usual disclaimer on what environment isolation entails.

> The map generator threads, which also contained the above mentioned Lua environment,

Reviewer: contained -> contain ?

the mapgen environment. After that only callbacks happen. The mapgen env
does not have a global step or timer.

* `minetest.register_mapgen_script(path)`:
    * Register a path to a Lua file to be imported when a mapgen environment
      is initialized. Run in order of registration.

### List of APIs exclusive to the mapgen env

* `minetest.register_on_generated(function(vmanip, minp, maxp, blockseed))`
    * Called after the engine mapgen finishes a chunk but before it is written to
      the map.
    * Chunk data resides in `vmanip`. Other parts of the map are not accessible.
      The area of the chunk if comprised of `minp` and `maxp`, note that is smaller
      than the emerged area of the VoxelManip.
      Note: calling `read_from_map()` or `write_to_map()` on the VoxelManipulator object
      is not necessary and is disallowed.
    * `blockseed`: 64-bit seed number used for this chunk
* `minetest.save_gen_notify(id, data)`
    * Saves data for retrieval using the gennotify mechanism (see [Mapgen objects]).
    * Data is bound to the chunk that is currently being processed, so this function
      only makes sense inside the `on_generated` callback.

>     * `id`: user-defined ID (a string)

Reviewer: newline not rendered  
"user-defined ID (a string) By convention these ..."

Author: I can't remember how to fix that, so I'll leave it :sweat_smile:

Reviewer: two trailing spaces result in a `<br>`

    * with all functions and userdata values replaced by `true`, calling any
      callbacks here is obviously not possible

Mapgen environment
------------------

The engine runs the map generator on separate threads, each of these also has
a Lua environment. Its primary purpose is to allow mods to operate on newly
generated parts of the map to e.g. generate custom structures.
Internally it is referred to as "emerge environment".

> Refer to the above section for the usual disclaimer on what environment isolation entails.

Reviewer 2: ```suggestion
Refer to [Async environment] for the usual disclaimer on what environment isolation entails.
```
